assembly/pom.xml

```diff
@@ -159,6 +169,78 @@
             </target>
           </configuration>
       </plugin>
+      <plugin>
+        <!--
+          Here we download ammonite dependency required for Spark Connect REPL and copy
+          Spark Connect client to target's jars/connect-repl directory. Both jars will
+          only be loaded when we run Spark Connect shell, see also
+          AbstractCommandBuilder.buildClassPath and SPARK-48936.
+        -->
+        <groupId>org.codehaus.mojo</groupId>
+        <artifactId>exec-maven-plugin</artifactId>
```

Comment:
Made a PR to prune unnecessary jars (https://github.com/apache/spark/pull/47705), still testing.